pkg/sidecarexec/sandbox_cmd_runner_linux.go

```diff
@@ -0,0 +1,282 @@
+// which sandbox has initiated within a new user/mount namespace.
+type SandboxWrap struct{}
+
+type mountPoint struct {
+	Path      string
+	File      bool
+	Writeable bool
+	NonMount  bool
+	Content   string
+}
+
+// ExecuteCmd mounts various directories needed to execute sandboxed command.
+// As a last step it exec-s into given command, replacing the current process.
+func (w SandboxWrap) ExecuteCmd(rawArgs []string) error {
+	runtime.LockOSThread() // as a safety precaution for syscalls below
+
+	args := newSandboxWrapArgs(rawArgs[0])
+
+	// Keep default environment as minimal as possible
+	mounts := []mountPoint{
+		{Path: "usr/bin"},
+		{Path: "tmp", NonMount: true},
+	}
+
+	for _, path := range args.VisiblePaths {
+		if !strings.HasPrefix(path, limitDir) {
+			return fmt.Errorf("Expected visible path to start with '%s', but was '%s'", limitDir, path)
+		}
+		mounts = append(mounts, mountPoint{Path: path[1:], Writeable: true})
+	}
+
+	if args.Posix {
+		// Example: calling out to git
+		mounts = append(mounts, []mountPoint{
+			{Path: "etc", NonMount: true},
+			{Path: "etc/passwd", File: true, NonMount: true, Content: "root:x:0:0:/home/kapp-controller:/usr/sbin/nologin\n"},
+			{Path: "bin"}, // e.g. bin/sh was referenced by git executables
+			{Path: "lib"},
+			{Path: "lib64"}, // some dynamic libraries
+			{Path: "var/lib"},
+			{Path: "usr/lib"},
+			{Path: "usr/libexec"}, // internal git binaries
+			{Path: "usr/share"},   // internal git data
+			{Path: "usr/sbin"},
+			// TODO get rid of 'dev/null'? (used by vendir today)
+			{Path: "dev", NonMount: true},
+			{Path: "dev/null", NonMount: true, File: true},
+		}...)
+	}
+
+	if args.Network {
+		mounts = append(mounts, []mountPoint{
+			// Cannot mount whole 'etc' directory directly (mount returns exit status 32)
+			// Mounting individual child files/directories makes it work.
+			{Path: "etc/pki"},
+			{Path: "etc/resolv.conf", File: true},
+			{Path: "etc/hostname", File: true},
+			{Path: "etc/hosts", File: true},
+		}...)
+	}
+
+	err := w.configureMounts(mounts, args.NewRootDir)
+	if err != nil {
+		return err
+	}
+
+	err = w.pivotRoot(args.NewRootDir)
+	if err != nil {
+		return err
+	}
+
+	// Not all command specify execution directory
+	if len(args.CmdDir) > 0 {
+		err = os.Chdir(args.CmdDir)
+		if err != nil {
+			return fmt.Errorf("Chdir '%s': %s", args.CmdDir, err)
+		}
+	}
+
+	return syscall.Exec(args.CmdPath, append([]string{args.CmdPath}, args.CmdArgs...), os.Environ())
+}
+
+func (SandboxWrap) configureMounts(mounts []mountPoint, newRootDir string) error {
+	for _, mount := range mounts {
+		if mount.File {
+			file, err := os.Create(filepath.Join(newRootDir, mount.Path))
+			if err != nil {
+				return fmt.Errorf("Creating mount file '%s': %s", mount.Path, err)
+			}
+			if len(mount.Content) > 0 {
+				_, err = file.Write([]byte(mount.Content))
+				if err != nil {
+					return fmt.Errorf("Writing into file '%s': %s", mount.Path, err)
```

Comment:
do we need to close the file if there's an error while writing?

Reply:
if we want to be good citizens (technically process dies so all fds are closed by the kernel). ill make the change.